arcgis-ios-sdk-samples/Layers/Export tiles/ExportTilesViewController.swift

```diff
@@ -17,154 +17,238 @@ import UIKit
+        // the tile cache will use the current compact version 2 format.
+        // See more in the doc of
+        // `AGSExportTileCacheTask.exportTileCacheJob(with:downloadFileURL:)`.
+        temporaryFolderURL.appendingPathComponent("myTileCache.\(fileFormat.fileExtension)")
+    }
+    
+    // MARK: Actions
+    
+    @IBAction func exportTilesBarButtonTapped(_ sender: UIBarButtonItem) {
+        if let exportJob = job {
+            // If it is downloading, cancel download.
+            exportJob.progress.cancel()
+            job = nil
+        } else {
+            // Otherwise, try to download when exporting tiles is allowed.
+            guard let mapServiceInfo = exportTask.mapServiceInfo, mapServiceInfo.exportTilesAllowed else { return }
+            // Create an action sheet to choose export format.
+            let alertController = UIAlertController(
+                title: "Choose Export Format",
+                message: nil,
+                preferredStyle: .actionSheet
+            )
+            let allowedFormats: [TilePackageFormat] = mapServiceInfo.exportTileCacheCompactV2Allowed ? [.tpk, .tpkx] : [.tpk]
+            allowedFormats.forEach { fileFormat in
+                alertController.addAction(
+                    UIAlertAction(title: fileFormat.description, style: .default) { [unowned self] _ in
+                        self.initiateDownload(exportTask: exportTask, downloadFileURL: self.getDownloadURL(fileFormat: fileFormat))
+                    }
+                )
+            }
+            let cancelAction = UIAlertAction(title: "Cancel", style: .cancel)
+            alertController.addAction(cancelAction)
+            alertController.popoverPresentationController?.barButtonItem = sender
+            present(alertController, animated: true)
+        }
+    }
+    
+    @IBAction func closeButtonTapped(_ sender: UIButton) {
+        visualEffectView.isHidden = true
+        // Release the map in order to free the tiled layer.
+        previewMapView.map = nil
+    }
+    
+    // MARK: UIViewController
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Add the source code button item to the right of navigation bar.
+        (navigationItem.rightBarButtonItem as! SourceCodeBarButtonItem).filenames = ["ExportTilesViewController"]
+        // Load the export task.
+        exportTask.load { [weak self] error in
+            guard let self = self else { return }
+            if let error = error {
+                self.presentAlert(error: error)
+            } else {
+                self.exportTilesBarButtonItem.isEnabled = true
             }
-            print("deleted all local data")
-        } catch {
-            print(error)
         }
     }
+    
+    deinit {
+        // Remove the temporary folder and all content in it.
+        try? FileManager.default.removeItem(at: temporaryFolderURL)
+    }
```

Comment:
Is it necessary to have both this and the `removeTemporaryTilePackages()` function?

Reply:
Yes. The `removeTemporaryTilePackages` function removes all the files in the folder, but doesn't remove the folder, so if a user want to export another tile package at a different viewpoint, he can save to disk without overriding existing files.
The remove in `deinit`, on the other hand, remove all the files including the folder as a clean-up.

Think about a case like this: user taps export at San Diego once, then closes the preview, then export at LA, and quit the app without closing the preview.

However, I agree that the `removeTemporaryTilePackages` function might be moved into the close action. Currently it is get called when new export is initiated. What do you think?

Reply:
Hm, yeah it might make more sense to move the function call to the close action.